### ISSUE TITLE
No 'Access-Control-Allow-Origin'

### ISSUE THREAD

**Issue by baftijarovskiA** [fruitcake/laravel-cors#421]
I write everything as stated in the readme.md file but still get an error on my requests.
_**No 'Access-Control-Allow-Origin' header is present on the requested resource.**_
I am using Laravel 6

**Comment by felixivance** [comment]
Experiencing the same problem, what might be the issue

**Comment by jekabsmilbrets** [comment]
Same here

**Comment by orijmm** [comment]
Same

**Comment by jm21** [comment]
GET requests seem to be working fine, POST requests are not going through in my case

**Comment by barryvdh** [comment]
What is your config?

**Comment by jm21** [comment]
````
return [

    /*
    |--------------------------------------------------------------------------
    | Laravel CORS Options
    |--------------------------------------------------------------------------
    |
    | The allowed_methods and allowed_headers options are case-insensitive.
    |
    | You don't need to provide both allowed_origins and allowed_origins_patterns.
    | If one of the strings passed matches, it is considered a valid origin.
    |
    | If array('*') is provided to allowed_methods, allowed_origins or allowed_headers
    | all methods / origins / headers are allowed.
    |
    */

    /*
     * You can enable CORS for 1 or multiple paths.
     * Example: ['api/*']
     */
    'paths' => ['api/*'],

    /*
    * Matches the request method. `[*]` allows all methods.
    */
    'allowed_methods' => ['*'],

    /*
     * Matches the request origin. `[*]` allows all origins.
     */
    'allowed_origins' => ['*'],

    /*
     * Matches the request origin with, similar to `Request::is()`
     */
    'allowed_origins_patterns' => ['*'],

    /*
     * Sets the Access-Control-Allow-Headers response header. `[*]` allows all headers.
     */
    'allowed_headers' => ['*'],

    /*
     * Sets the Access-Control-Expose-Headers response header.
     */
    'exposed_headers' => false,

    /*
     * Sets the Access-Control-Max-Age response header.
     */
    'max_age' => false,

    /*
     * Sets the Access-Control-Allow-Credentials header.
     */
    'supports_credentials' => false,
];
````

**Comment by barryvdh** [comment]
Did you clear the config cache? What is the actual path you post to?

**Comment by jm21** [comment]
I've cleared config cache. Posting to http://localhost/capslock-admin/backend/public/api/login

**Comment by barryvdh** [comment]
That's an unusual path, perhaps try * or *api*

**Comment by jm21** [comment]
If I set paths to  'paths' => ['*'] GET requests work.  POST requests are not working at all

**Comment by barryvdh** [comment]
And you added the middleware to the global middleware? Not just the group?

**Comment by georgenis** [comment]
same problem here, I did add the HandleCors to the globale middlewares...set config to '*' on neccessary fields...

**Comment by jm21** [comment]
I added the middleware, didn't work. Removed it, still didn't work

**Comment by barryvdh** [comment]
Cam you add some logging to https://github.com/fruitcake/laravel-cors/blob/master/src/HandleCors.php ? Maybe check if shouldRun or isMatchingPath are returning true correctly.

**Comment by georgenis** [comment]
I found my issue: I had to sent "Accept: application/json" HTTP-Header to laravel...without "accept"-header it does not work.

**Comment by phuchnh** [comment]
Maybe you should check `\Fruitcake\Cors\HandleCors::class` already added in `app/Http/Kernel.php` or not?

**Comment by shadowgroundz** [comment]
+1 I have some issue to. No problem with get request

**Comment by suppergohan** [comment]
I had to sent "Accept: application/json" HTTP-Header to laravel. Not working.

**Comment by shadowgroundz** [comment]
In my case, it happen if I read base64 data from $request. But if I not read there is no problem

**Comment by barryvdh** [comment]
Are you not having errors (eg. CSRF errors?) Can you put some loggers? Is the same request working with Postman? Can you show how you output/create the response?

**Comment by shadowgroundz** [comment]
> Are you not having errors (eg. CSRF errors?) Can you put some loggers? Is the same request working with Postman? Can you show how you output/create the response?

Hi, thanks for your response. This my config

<?php

return [

```
    /*
    |--------------------------------------------------------------------------
    | Laravel CORS Options
    |--------------------------------------------------------------------------
    |
    | The allowed_methods and allowed_headers options are case-insensitive.
    |
    | You don't need to provide both allowed_origins and allowed_origins_patterns.
    | If one of the strings passed matches, it is considered a valid origin.
    |
    | If array('*') is provided to allowed_methods, allowed_origins or allowed_headers
    | all methods / origins / headers are allowed.
    |
    */

    /*
     * You can enable CORS for 1 or multiple paths.
     * Example: ['api/*']
     */
    'paths' => ['api/*'],

    /*
    * Matches the request method. `[*]` allows all methods.
    */
    'allowed_methods' => ['*'],

    /*
     * Matches the request origin. `[*]` allows all origins.
     */
    'allowed_origins' => ['*'],

    /*
     * Matches the request origin with, similar to `Request::is()`
     */
    'allowed_origins_patterns' => [],

    /*
     * Sets the Access-Control-Allow-Headers response header. `[*]` allows all headers.
     */
    'allowed_headers' => ['*'],

    /*
     * Sets the Access-Control-Expose-Headers response header.
     */
    'exposed_headers' => false,

    /*
     * Sets the Access-Control-Max-Age response header.
     */
    'max_age' => false,

    /*
     * Sets the Access-Control-Allow-Credentials header.
     */
    'supports_credentials' => true,
];
```

When I hit use my ReactJs app respose is like this
<img width="758" alt="Screen Shot 2020-02-24 at 15 33 19" src="https://user-images.githubusercontent.com/17138664/75137872-5b8e3000-571b-11ea-8eed-851bd75429d6.png">

But when I hit using postman result show what I want
<img width="1680" alt="Screen Shot 2020-02-24 at 15 36 58" src="https://user-images.githubusercontent.com/17138664/75137937-84aec080-571b-11ea-8e80-cf87654b2b55.png">

**Comment by barryvdh** [comment]
Are all of you using some sub-path? eg. localhost/yourproject/public/api.. ?

**Comment by shadowgroundz** [comment]
I use ***/public/*** just only for development

**Comment by nalletje** [comment]
> > Are you not having errors (eg. CSRF errors?) Can you put some loggers? Is the same request working with Postman? Can you show how you output/create the response?
> 
> Hi, thanks for your response. This my config
> 
> ```
> <?php
> return [
>     'paths' => ['api/*'],
>     'allowed_methods' => ['*'],
>     'allowed_origins' => ['*'],
>     'allowed_origins_patterns' => [],
>     'allowed_headers' => ['*'],
>     'exposed_headers' => false,
>     'max_age' => false,
>     'supports_credentials' => true,
> ];
> ```
> 
> When I hit use my ReactJs app respose is like this
> <img alt="Screen Shot 2020-02-24 at 15 33 19" width="758" src="https://user-images.githubusercontent.com/17138664/75137872-5b8e3000-571b-11ea-8eed-851bd75429d6.png">
> 
> But when I hit using postman result show what I want
> <img alt="Screen Shot 2020-02-24 at 15 36 58" width="1680" src="https://user-images.githubusercontent.com/17138664/75137937-84aec080-571b-11ea-8e80-cf87654b2b55.png">

Are you sure the ReactJS is not sending a HEAD request before the POST request? Every time struggling with that again with NuxtJS. 

Mostly the server configuration needs to change to allow method request HEAD to fix this for me.

**Comment by shadowgroundz** [comment]
> > > Are you not having errors (eg. CSRF errors?) Can you put some loggers? Is the same request working with Postman? Can you show how you output/create the response?
> > 
> > 
> > Hi, thanks for your response. This my config
> > ```
> > <?php
> > return [
> >     'paths' => ['api/*'],
> >     'allowed_methods' => ['*'],
> >     'allowed_origins' => ['*'],
> >     'allowed_origins_patterns' => [],
> >     'allowed_headers' => ['*'],
> >     'exposed_headers' => false,
> >     'max_age' => false,
> >     'supports_credentials' => true,
> > ];
> > ```
> > 
> > 
> > When I hit use my ReactJs app respose is like this
> > <img alt="Screen Shot 2020-02-24 at 15 33 19" width="758" src="https://user-images.githubusercontent.com/17138664/75137872-5b8e3000-571b-11ea-8eed-851bd75429d6.png">
> > But when I hit using postman result show what I want
> > <img alt="Screen Shot 2020-02-24 at 15 36 58" width="1680" src="https://user-images.githubusercontent.com/17138664/75137937-84aec080-571b-11ea-8e80-cf87654b2b55.png">
> 
> Are you sure the ReactJS is not sending a OPTIONS request before the POST request? Every time struggling with that again with NuxtJS.
> 
> Mostly the server configuration needs to change to allow method request OPTIONS to fix this for me.

Yes sure. Not send option before hit any api. But not problem for any GET request

**Comment by miharbi** [comment]
Same error here

**Comment by sinceow** [comment]
I faced the same problem, and tried everything mentioned above, still not working

**Comment by nalletje** [comment]
Did you guys consider trying returning JSON (or checking if it returns JSON)?
In the postman screenshot i see he's returning an array instead of JSON.

**Comment by jeherrera** [comment]
I had this same issue and it seems to have been resolved by adding HandleCors::class to the top of the $middlewarePriority list in addition to the $middleware list as noted in the docs:

```
protected $middlewarePriority = [
        \Fruitcake\Cors\HandleCors::class, //<- this seems to have been the missing piece for me
        \Illuminate\Session\Middleware\StartSession::class,
        \Illuminate\View\Middleware\ShareErrorsFromSession::class,
        \App\Http\Middleware\Authenticate::class,
        \Illuminate\Routing\Middleware\ThrottleRequests::class,
        \Illuminate\Session\Middleware\AuthenticateSession::class,
        \Illuminate\Routing\Middleware\SubstituteBindings::class,
        \Illuminate\Auth\Middleware\Authorize::class,
    ];
```

**Comment by p2sousa** [comment]
I had the same problem, but I was registering the `middleware` only in the `api group`, it was resolved by registering it in the `global middleware`.

`\App\Http\Kernel.php`
```php
protected $middleware = [
        \App\Http\Middleware\TrustProxies::class,
        \App\Http\Middleware\CheckForMaintenanceMode::class,
        \Illuminate\Foundation\Http\Middleware\ValidatePostSize::class,
        \App\Http\Middleware\TrimStrings::class,
        \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull::class,
        \Fruitcake\Cors\HandleCors::class,
];
```

in `config/cors.php` I had to specify the path.

`config/cors.php`
```php
$origins = env('CORS_ORIGINS', []);
return [
    'paths' => ['api/*'],
    'allowed_methods' => ['*'],
    'allowed_origins' => explode(',', $origins),
    'allowed_origins_patterns' => [],
    'allowed_headers' => ['*'],
    'exposed_headers' => false,
    'max_age' => false,
    'supports_credentials' => false,
];
```

**Comment by commeco** [comment]
> I had this same issue and it seems to have been resolved by adding HandleCors::class to the top of the $middlewarePriority list in addition to the $middleware list as noted in the docs:
> 
> ```
> protected $middlewarePriority = [
>         \Fruitcake\Cors\HandleCors::class, //<- this seems to have been the missing piece for me
>         \Illuminate\Session\Middleware\StartSession::class,
>         \Illuminate\View\Middleware\ShareErrorsFromSession::class,
>         \App\Http\Middleware\Authenticate::class,
>         \Illuminate\Routing\Middleware\ThrottleRequests::class,
>         \Illuminate\Session\Middleware\AuthenticateSession::class,
>         \Illuminate\Routing\Middleware\SubstituteBindings::class,
>         \Illuminate\Auth\Middleware\Authorize::class,
>     ];
> ```

Hi,
in our case the sort order in the $middleware array was the solution.
    
```
protected $middleware = [
        \Fruitcake\Cors\HandleCors::class,         <-- It must be the first entry here!
        \App\Http\Middleware\TrustProxies::class,
        \App\Http\Middleware\CheckForMaintenanceMode::class,
        \Illuminate\Foundation\Http\Middleware\ValidatePostSize::class,
        \App\Http\Middleware\TrimStrings::class,
        \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull::class,
    ];
```

**Comment by barryvdh** [comment]
Can you check which middleware makes the difference?

**Comment by commeco** [comment]
> Can you check which middleware makes the difference?

Sorry Barry. It was a false alarm. The problem was the php opcache. :-/

**Comment by ahmedkhan847** [comment]
@barryvdh  I am facing the same issue with the CORS tried everything its not working

**Comment by alexanderbenerink** [comment]
> @barryvdh I am facing the same issue with the CORS tried everything its not working

Me too. I can't wrap my head around where the problem originates from. I created a clean laravel 6 installation, merged our old projects files into it, fixed the bugs and added this package but to no avail. Locally it works, mind you. It also works when I send my headers to a script in JSFiddle. My api is then visible and working. 

But when it's on production/server the websites connected to our api still get this error. I changed permissions for storage, cleared all the caches/config, 
I use fruitcake/laravel-cors: "^1.0", my laravel version is 6.18.0, my dev server php version is 7.4 and my project is php version 7.2.9.

If I add this in or outside the "IfModule mod_headers.c" tag in my .htaccess file:

>Header always set Access-Control-Allow-Origin " * "
>Header always set Access-Control-Allow-Methods: " * ":

It changes the error from this:

> "Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

to this:

> "Response to preflight request doesn't pass access control check: It does not have HTTP ok status.""

I feel like I'm close, however that still won't get me anywhere. The HandleCors class is inside the global $middleware, inside the Cors config file I use a wildcard for allowed_origins, allowed_methods and allowed_headers. Changing the path won't have any effect but it is "api/*".

Would it possibly have something to do with my api routes file? 

routes > api.php:

```
<?php

use Illuminate\Http\Request;

/*
|--------------------------------------------------------------------------
| API Routes
|--------------------------------------------------------------------------
|
| Here is where you can register API routes for your application. These
| routes are loaded by the RouteServiceProvider within a group which
| is assigned the "api" middleware group. Enjoy building your API!
|
*/

Route::middleware('auth:api')->get('/user', function (Request $request) {
    return $request->user();
});

//Route::group(['config' => 'cors'], function () {
    Route::options('{any}'); // CORS Preflight

    Route::get('/app/{application}/layout', 'Api\LayoutController@index');
    Route::get('/app/{application}', 'Api\PageController@index');
    Route::get('/app/{application}/{page_id}', 'Api\PageController@show');
    Route::post('/app/{application}/search', 'Api\SearchController@index');
    Route::post('/app/{application}/review', 'Api\ReviewController@index');
    Route::post('/app/{application}/review-comment', 'Api\ReviewController@comment');
    Route::post('/app/{application}/form', 'Api\FormController@index');
    Route::post('/app/{application}/message', 'Api\MessageController@store');
    Route::post('/app/{application}/chat', 'Api\MessageController@index');
//});
```
Or has it something to do with the server? And adding the mod_headers.c to my httpd.conf on the server won't give me any results either. 
I've been at this for a week now.

**Comment by barryvdh** [comment]
You don;t have to define the Route::options and not sure what the config => cors is?

**Comment by alexanderbenerink** [comment]
> You don;t have to define the Route::options and not sure what the config => cors is?

Originally we had a custom middleware but it didn't work anymore so we switched to the package. I uncommented it but I should've removed it for clarity in my previous post. sorry

**Comment by barryvdh** [comment]
But what is  Route::options('{any}');?

**Comment by alexanderbenerink** [comment]
> But what is Route::options('{any}');?

To allow preflighting for every route/slug/whathaveyou. I ported the project from 5.4 to 6. That might have something to do with it? I doubt it

**Comment by IceBay** [comment]
In my case, set `supports_credentials` to `true`, it works

**Comment by IceBay** [comment]
> In my case, set `supports_credentials` to `true`, it works

Because i set credentials in React.js

**Comment by Bluefitdev** [comment]
It doesn't seem to work on mine either. 

Sorted the middleware class, put everything on * in config file, nothing seems to work....

**Comment by Bluefitdev** [comment]
Mine is having an error on the `addActualHeaderRequest`

```
    public function handle($request, Closure $next)
    {
        // Check if we're dealing with CORS and if we should handle it
        if (!$this->shouldRun($request)) {
            Log::info('Goes in should run');

            return $next($request);
        }

        // For Preflight, return the Preflight response
        if ($this->cors->isPreflightRequest($request)) {
            return $this->cors->handlePreflightRequest($request);
        }

        // If the request is not allowed, return 403
        if (!$this->cors->isActualRequestAllowed($request)) {
            Log::info('Goes in not allowed');

            return new Response('Not allowed in CORS policy.', 403);
        }

        // Handle the request
        $response = $next($request);
        Log::info('Handled is: '.$response);

        // Add the CORS headers to the Response
        return $this->addHeaders($request, $response);
    }
```

Response returns the correct one in the laravel.log

But on the implode in CorsService it has error

```
[2020-03-11 01:03:53] local.ERROR: implode(): Invalid arguments passed {"userId":1,"exception":"[object] (ErrorException(code: 0): implode(): Invalid arguments passed at /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/asm89/stack-cors/src/Asm89/Stack/CorsService.php:94)
[stacktrace]
#0 [internal function]: Illuminate\\Foundation\\Bootstrap\\HandleExceptions->handleError(2, 'implode(): Inva...', '/Users/bluefitm...', 94, Array)
#1 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/asm89/stack-cors/src/Asm89/Stack/CorsService.php(94): implode(', ', true)
#2 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/fruitcake/laravel-cors/src/HandleCors.php(120): Asm89\\Stack\\CorsService->addActualRequestHeaders(Object(Illuminate\\Http\\JsonResponse), Object(Illuminate\\Http\\Request))
#3 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/fruitcake/laravel-cors/src/HandleCors.php(64): Fruitcake\\Cors\\HandleCors->addHeaders(Object(Illuminate\\Http\\Request), Object(Illuminate\\Http\\JsonResponse))
#4 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fruitcake\\Cors\\HandleCors->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#5 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(41): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#6 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#7 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Routing/Middleware/ThrottleRequests.php(59): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#8 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Routing\\Middleware\\ThrottleRequests->handle(Object(Illuminate\\Http\\Request), Object(Closure), 60, '1')
#9 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#10 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Routing/Router.php(683): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#11 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Routing/Router.php(658): Illuminate\\Routing\\Router->runRouteWithinStack(Object(Illuminate\\Routing\\Route), Object(Illuminate\\Http\\Request))
#12 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Routing/Router.php(624): Illuminate\\Routing\\Router->runRoute(Object(Illuminate\\Http\\Request), Object(Illuminate\\Routing\\Route))
#13 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Routing/Router.php(613): Illuminate\\Routing\\Router->dispatchToRoute(Object(Illuminate\\Http\\Request))
#14 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(165): Illuminate\\Routing\\Router->dispatch(Object(Illuminate\\Http\\Request))
#15 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}(Object(Illuminate\\Http\\Request))
#16 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#17 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#18 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#19 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#20 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#21 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#22 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(63): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#23 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\CheckForMaintenanceMode->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#24 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/fruitcake/laravel-cors/src/HandleCors.php(60): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#25 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fruitcake\\Cors\\HandleCors->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#26 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#27 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fideloper\\Proxy\\TrustProxies->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#28 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#29 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(140): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#30 /Users/bluefitmarketing/Valet-Sites/salt-backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(109): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter(Object(Illuminate\\Http\\Request))
#31 /Users/bluefitmarketing/Valet-Sites/salt-backend/public/index.php(55): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))
#32 /Users/bluefitmarketing/.composer/vendor/laravel/valet/server.php(158): require('/Users/bluefitm...')
#33 {main}
"} 
[2020-03-11 01:03:53] local.INFO: Handled is: HTTP/1.1 500 Internal Server Error
Cache-Control:         no-cache, private
Content-Type:          application/json
Date:                  Wed, 11 Mar 2020 01:03:53 GMT
X-Ratelimit-Limit:     60
X-Ratelimit-Remaining: 58
```

What's happening?

**Comment by Bluefitdev** [comment]
I'm so dumb, apparently it brings down to the 

```
        if ($this->options['exposedHeaders']) {
            $response->headers->set('Access-Control-Expose-Headers', implode(', ', $this->options['exposedHeaders']));
        }
```

I set the config as true, didn't really know what it meant.

Set it to false after debugging now its working

Thanks!

**Comment by barryvdh** [comment]
I've updated the defaults here to make it more clear: https://github.com/fruitcake/laravel-cors/commit/6d46adb112cd844851b81968b8b55b3d12abc43a

Did anyone else have the same problem or debug something?

**Comment by arjan-kwikqy** [comment]
Hi,

Having the same problem here. Everything works fine except when Im adding a custom header called 'workspace' to the requests.. 

Config:

`'paths' => ['api/*'],
    'allowed_methods' => ['*'],
    'allowed_origins' => ['*'],
    'allowed_origins_patterns' => [],
    'allowed_headers' => ['*'],
    'exposed_headers' => [],
    'max_age' => 0,
    'supports_credentials' => false`

**Comment by barryvdh** [comment]
I've added some tests here: https://github.com/fruitcake/laravel-cors/commit/70b790c92d9a1a0942caf21cfd94d5fa50ce6659
It should work with a wildcard. 

Does it work when adding the head manually instead of the wildcard?

**Comment by barryvdh** [comment]
> Hi,
> 
> Having the same problem here. Everything works fine except when Im adding a custom header called 'workspace' to the requests..
> 
> Config:
> 
> `'paths' => ['api/*'], 'allowed_methods' => ['*'], 'allowed_origins' => ['*'], 'allowed_origins_patterns' => [], 'allowed_headers' => ['*'], 'exposed_headers' => [], 'max_age' => 0, 'supports_credentials' => false`
> 
> Using Laravel 7

Did you clear your config? Can you post the exact Request headers as sent by the browser? And the response headers.

Does the Options request succeed? What is the exact error+response?

**Comment by wassim** [comment]
@barryvdh Having the same problem here. Using the default config on production and doing a request with axios:

```
await axios.post(`${process.env.API_URL}/handle`, { data }, {
                headers: {
                    Authorization: 'Bearer ' + process.env.API_TOKEN,
                    Accept: 'application/json',
                }
            })
```

`Access to XMLHttpRequest at 'hidden' from origin 'hidden' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

**Comment by barryvdh** [comment]
Did you allow credentials?

**Comment by wassim** [comment]
@barryvdh you mean 'supports_credentials'? It is set to true.

**Comment by barryvdh** [comment]
so it's not the default config? Can you please post the exact config + request headers and response headers (as seen in dev tools)

**Comment by wassim** [comment]
@barryvdh 

**Cors.php**

```
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Cross-Origin Resource Sharing (CORS) Configuration
    |--------------------------------------------------------------------------
    |
    | Here you may configure your settings for cross-origin resource sharing
    | or "CORS". This determines what cross-origin operations may execute
    | in web browsers. You are free to adjust these settings as needed.
    |
    | To learn more: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
    |
    */

    'paths' => ['api/*'],

    'allowed_methods' => ['*'],

    'allowed_origins' => ['*'],

    'allowed_origins_patterns' => [],

    'allowed_headers' => ['*'],

    'exposed_headers' => false,

    'max_age' => false,

    'supports_credentials' => true,

];
```

**Request headers**

```
Accept: application/json
Authorization: Bearer [hidden_api_token]
Content-Type: application/json;charset=UTF-8
DNT: 1
Referer: https://mywebsite.com/
Sec-Fetch-Dest: empty
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36
```

**Response**

```
Access to XMLHttpRequest at 'https://backend.mywebsite.com/handle' from origin 'https://mywebsite.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
index.js:35 Error: Network Error
    at LYNF.e.exports (createError.js:17)
    at XMLHttpRequest.p.onerror (xhr.js:91)
xhr.js:183 POST https://backend.mywebsite.com/handle net::ERR_FAILED
```

**Comment by barryvdh** [comment]
I've created an issue template, can you all verify these steps and create a new issue if it doesn't work? 

**Before you start**
[ ] Update to the latest version by running `composer update fruitcake/laravel-cors`

**Check your config**
[ ]  Double-check your config file with the version from the repo. Make sure the `paths` property is correctly set. Start by allowing as much as possible.
[ ]  Make sure the middleware is added to the global middleware in your Http Kernel (not group)

**Clear your caches**
Please do these steps again before submitting an issue:
[ ]  Clear your config cache `php artisan config:clear`, route cache (`php artisan route:clear`) and normal cache (`php artisan cache:clear`).
[ ]  Make sure your permissions are setup correctly (eg. storage is writable)

**Comment by yanlinaung30** [comment]
In my case, there is no problem when api request to `api/product` is made. But there is  error `No 'Access-Control-Allow-Origin'`  when I try to access `api/products/1`. I am using Laravel 7 with config/cors.php config.

**Comment by yanlinaung30** [comment]
It is solved when I delete the code in controller :3 . So please delete if you use.
```
dd($product);
```

**Comment by saavedraphp** [comment]
please I have the problem 3 days and I can't find a solution "Laravel 6" that frustrating

Having the same problem here. Using the default config on production and doing a request with 

             axios.get(`http://localhost/estados/pais/${this.selected_pais}`).then((response) => {                
                    this.careers = response.data;
                })
 
-----------------
 protected $middleware = [
        \Fruitcake\Cors\HandleCors::class,
-------------------

    'paths'                    => ['api/*'],

    'allowed_methods'          => ['*'],

    'allowed_origins'          => ['*'],

    'allowed_origins_patterns' => [],

    'allowed_headers'          => ['*'],

    'exposed_headers'          => [],

    'max_age'                  => 0,

    'supports_credentials'     => false,

![error1](https://user-images.githubusercontent.com/18977975/80052114-7f81ae00-84df-11ea-875a-5bb7429754ea.png)

**Comment by afiqiqmal** [comment]
For my case, After several hours drill down to HandlerCors and CorsService class, the path should be set, else all request will become non-cors

**Comment by barryvdh** [comment]
@saavedraphp you allow api/* but your request is /estados ?

**Comment by carlomigueldy** [comment]
is there a fix for this issue yet?

**Comment by Brandutchmen** [comment]
@carlomigueldy What I did to make this work is code in endpoints within Laravel to use and then use a Guzzle request within Laravel to request the external source.

**Comment by carlomigueldy** [comment]
can u show an example? @Brandutchmen

**Comment by jardila** [comment]
> For my case, After several hours drill down to HandlerCors and CorsService class, the path should be set, else all request will become non-cors

I have the same problem. in my case, i have the route 'api/carga/detail/{id}' with ok response, but the route 'api/carga/uploader' return "No 'Access-Control-Allow-Origin'" error. the only difference in the last one endpoint is that it comes with form-data (files) from the origin.

My cors setting is:

![image](https://user-images.githubusercontent.com/3129156/80223986-10a77080-8617-11ea-8ed0-260ab89decb3.png)



![image](https://user-images.githubusercontent.com/3129156/80223797-c2926d00-8616-11ea-9320-05551a1ec617.png)


I hope to find any solution.

**Comment by Brandutchmen** [comment]
@jardila It looks like your api request url is your yoururl/api/cargo/uploader, but your devtools is using port 3000. That might be contributing to your problem. Otherwise, I assume that route is resolving to a controller in Laravel? What is your config on that one look like?

**Comment by jardila** [comment]
> @jardila It looks like your api request url is your yoururl/api/cargo/uploader, but your devtools is using port 3000. That might be contributing to your problem. Otherwise, I assume that route is resolving to a controller in Laravel? What does your controller on that one look like?

Thanks @Brandutchmen for your feedback, yes, my app is using the 3000 port but that is not the problem, searching a bit, i find that the problem might be nginx because i had forgotten set '**client_max_body_size**' variable (i'm using files very long) and when i put  the corresponding value to nginx '**client_max_body_size**'  the error now is the request send my FormData empty and my controller receive it but generate an error because the file form does not come with the request.

I will find any solution and will post here if i can do it.

**Comment by Brandutchmen** [comment]
@carlomigueldy So here is an sample of a working CORS setup on one of my projects:
Controller:
```
  public function getData(Request $request)
    {
        $client = new Client();
        $response = $client->request('GET', 'https://jsonplaceholder.typicode.com/users/1'); // Any URL you want.
        $statusCode = $response->getStatusCode();
        $body = $response->getBody()->getContents();
        return $body;
    }
```

Route (routes /getData to the controller@getData) 

Cors config: (This is probably not configured in the most optimal way, just a quick test)
```
    'paths' => ['*'],

    /*
    * Matches the request method. `[*]` allows all methods.
    */
    'allowed_methods' => ['*'],

    /*
     * Matches the request origin. `[*]` allows all origins.
     */
    'allowed_origins' => ['*'],

    /*
     * Matches the request origin with, similar to `Request::is()`
     */
    'allowed_origins_patterns' => ['*'],

    /*
     * Sets the Access-Control-Allow-Headers response header. `[*]` allows all headers.
     */
    'allowed_headers' => [
        '*',
        'Content-Type',
        'X-Auth-Token',
        'Origin',
        'Authorization',
        'Access-Control-Allow-Origin',
    ],

    /*
     * Sets the Access-Control-Expose-Headers response header.
     */
    'exposed_headers' => false,

    /*
     * Sets the Access-Control-Max-Age response header.
     */
    'max_age' => false,

    /*
     * Sets the Access-Control-Allow-Credentials header.
     */
    'supports_credentials' => false,
];
```

Vue Method:
```
        axios
            .get('/getData')
                .then(function (response) {
                    // handle success
                     console.log(response.data);
                })
        },
```

**Comment by Brandutchmen** [comment]
@jardila Yeah. That does sound like it could be affecting it. I haven't seen client_max_body_size cause cors errors before. Interesting.

**Comment by jardila** [comment]
> @jardila Yeah. That does sound like it could be affecting it. I haven't seen client_max_body_size cause cors errors before. Interesting.

Effectively, the problem was in my nginx and php configuration. When i configure my nginx and php upload file size in order to allow upload files very long the error disappears. Now the error is in another process but it has nothing to do with cors configuration.

This was in my nginx log error file:

> 2020/04/24 12:41:40 [error] 18894#18894: *22061 **client intended to send too large body**: 5874328 bytes, client: 127.0.0.1, server: cobranza.test, request: "POST /api/carga/uploader HTTP/1.1", host: "cobranza.test", referrer: "http://cobranza.test:3000/app/gestion-clientes/cargas/deuda/1"

Thank you so much.

**Comment by jardila** [comment]
Please, review this for fix any problem with files in the request.
https://www.cyberciti.biz/faq/fix-client-intended-to-send-too-large-body-xyz-bytes-in-nginx/

**Comment by FuriosoJack** [comment]
> ¿Sucede con php artisan: serve? ¿Borró el caché (config, etc.) y verificó los permisos para las carpetas de almacenamiento / bootstrap?
in case it works with php artisan serve?
what could be done. Thank you

**Comment by Punksolid** [comment]
Before you start
I had a setup working with laravel 5.8 but it did a manual upgrade until 7.x, it used to work with 
**Barryvdh\Cors\HandleCors;**

But I changed it to use this following the upgrade guide of laravel
[ ] fruitcake/laravel-cors (v1.0.5 => v1.0.6)

Check your config
My config is set this way
```php
    'paths' => ['api/*'],
    'allowed_methods' => ['*'],
    'allowed_origins' => ['*'],
    'allowed_origins_patterns' => [],
    'allowed_headers' => ['*'],
    'exposed_headers' => [],
    'max_age' => 0,
    'supports_credentials' => false,
```
[ ] Make sure the middleware is added to the global middleware in your Http Kernel (not group)
Yes it is added in the global

Clear your caches
executed `php artisan cache:clear`, `php artisan config:clear` and `php artisan route:clear`
and chmod -R 777 /storage just for this environment

I saw you were asking to try this with php artisan serve, the problem persists within inside laradock and php artisan serve.

The route im trying to access is 
```Request URL:http://192.168.1.73:9527/api/sysadmin/v1/login```

These are the Request headers
```
Host: 192.168.1.73:9527
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:76.0) Gecko/20100101 Firefox/76.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Access-Control-Request-Method: POST
Access-Control-Request-Headers: content-type
Referer: http://localhost:9527/
Origin: http://localhost:9527
DNT: 1
Connection: keep-alive
```
And this the Response headers
```
TTP/1.1 200 OK
X-Powered-By: Express
Allow: GET,HEAD
Content-Type: text/html; charset=utf-8
Content-Length: 8
ETag: W/"8-ZRAf8oNBS3Bjb/SU2GYZCmbtmXg"
Date: Thu, 30 Apr 2020 01:19:32 GMT
Connection: keep-alive
```
This is for the OPTIONS verb during the preflight.

I will add a custom middleware for now. I hope this info helps.

**Comment by barryvdh** [comment]
So what is X-Powered-By: Express? Is that a Laravel server or expressjs? The endpoint needs the headers.

And please create separate issues.

**Comment by saavedraphp** [comment]
I found  the solution, the problem was how i put the address. thank you very much.

here the correct
 
Route::get('estados/pais/', 'UsuarioController@getEstadosByPais');

if (this.selected_pais !="") {
                axios.get(`http://127.0.0.1:80/estados/pais`, 
                {params: {pais_id: this.selected_pais} }).then((response) => {
                this.estados = response.data;
                document.getElementById('estado').disabled =false;

                });

SEE THE ERROR IMAGE ABOVE

**Comment by zawadsoon** [comment]
I have better one.
 
I had exact issue with cors. Everything was configured properly, urls was fine I was hitting right endpoint. But in middle of developemnt this cors error occured in console. 

What i did was permission change on storage/ directory.

I don't know how permissions impact cors settings, but after setting right permissins everyting is fine. It can be specific to my project so if I whenever investigate this I will share my experiences.

So, if you have this issue give a try to chmod 777 on storeage/ directory

**Comment by abiradak** [comment]
HeY Barry I am facing this only in Firefox :(
Here Is my Setup:
```
protected $middleware = [
        \App\Http\Middleware\TrustProxies::class,
        \App\Http\Middleware\CheckForMaintenanceMode::class,
        \Illuminate\Foundation\Http\Middleware\ValidatePostSize::class,
        \App\Http\Middleware\TrimStrings::class,
        \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull::class,
        \Fruitcake\Cors\HandleCors::class,
    ];
```

**Comment by barryvdh** [comment]
> And please create separate issues.
